src/tree.ts

```diff
@@ -427,30 +428,81 @@ class ProjectNode extends BaseNode {
   }
 }
 
-export class ProjectOutlineProvider implements vscode.TreeDataProvider<BaseNode> {
-  private readonly _changeEvent = new vscode.EventEmitter<BaseNode|null>();
-  get onDidChangeTreeData() { return this._changeEvent.event; }
+interface ExternalUpdateContext {
+  launchTargetName: string|null;
+  defaultTarget?: string;
+}
 
+class WorkspaceFolderNode extends BaseNode {
+  constructor(readonly wsFolder: vscode.WorkspaceFolder) { super(`wsf/${wsFolder.name}`); }
   private _children: BaseNode[] = [];
 
-  private _codeModel: cms.CodeModelContent = {configurations: []};
+  getOrderTuple() { return [this.id]; }
 
-  get codeModel() { return this._codeModel; }
+  getTreeItem() {
+    const item = new vscode.TreeItem(this.wsFolder.name, vscode.TreeItemCollapsibleState.Expanded);
+    item.iconPath = vscode.ThemeIcon.Folder;
+    item.label += ' — Workspace Folder';
+    return item;
+  }
 
-  updateCodeModel(model: cms.CodeModelContent|null, exCtx: {launchTargetName: string|null, defaultTargetName: string}) {
+  private _codeModel: cms.CodeModelContent = {configurations: []};
+  get codeModel() { return this._codeModel; }
+  updateCodeModel(model: cms.CodeModelContent|null, ctx: TreeUpdateContext) {
     if (!model || model.configurations.length < 1) {
+      this._children = [];
+      ctx.nodesToUpdate.push(this);
       return;
     }
     this._codeModel = model;
     const config = model.configurations[0];
-    const updates: BaseNode[] = [];
     const new_children: BaseNode[] = [];
     for (const pr of config.projects) {
-      const item = new ProjectNode(pr.name);
-      item.update(pr, {...exCtx, nodesToUpdate: updates});
+      const item = new ProjectNode(pr.name, ctx.folder);
+      item.update(pr, ctx);
       new_children.push(item);
     }
     this._children = new_children;
+  }
+
+  getChildren() {
+    return this._children;
+  }
+}
+
+export class ProjectOutlineProvider implements vscode.TreeDataProvider<BaseNode> {
+  private readonly _changeEvent = new vscode.EventEmitter<BaseNode|null>();
+  get onDidChangeTreeData() { return this._changeEvent.event; }
+
+  private readonly _folders = new Map<string, WorkspaceFolderNode>();
+
+  addAllCurrentFolders() {
+    for (const wsf of vscode.workspace.workspaceFolders || []) {
+      this._folders.set(wsf.name, new WorkspaceFolderNode(wsf));
+    }
+  }
+
+  addFolder(folder: vscode.WorkspaceFolder) {
+    this._folders.set(folder.name, new WorkspaceFolderNode(folder));
+    this._changeEvent.fire(null);
+  }
+
+  removeFolder(folder: vscode.WorkspaceFolder) {
+    this._folders.delete(folder.name);
+    this._changeEvent.fire(null);
+  }
+
+  updateCodeModel(folder: vscode.WorkspaceFolder, model: cms.CodeModelContent|null, ctx: ExternalUpdateContext) {
+    let existing = this._folders.get(folder.name);
+    if (!existing) {
+      rollbar.error('Updating code model on folder that does not yet exist?');
```

Comment:
Should this be localized?

Reply:
I don't think it would be helpful. But we can if you feel strongly.

Reply:
If it is an error that developers should see, then it needs to be localized.  If it is not an error and/or doesn't need to be seen, then use `log.debug`.  Either way, we've been localizing everything except for "trace" messages.